File: vrf-verify.js
const crypto = require('crypto');
const EC = require('elliptic').ec;
const BN = require('bn.js');

// secp256k1 曲线
const ec = new EC('secp256k1');
const curve = ec.curve;
const G = ec.g;
const n = ec.n;               // BN
const p = curve.p;            // BN

// ---------- 基础工具 ----------
function intToBytes(x, length = 32) {
    return x.toArrayLike(Buffer, 'be', length);
}

function sha256(data) {
    return crypto.createHash('sha256').update(data).digest();
}

function pointToBytes(P) {
    const xBuf = intToBytes(P.getX(), 32);
    const yBuf = intToBytes(P.getY(), 32);
    return Buffer.concat([Buffer.from([0x04]), xBuf, yBuf]);
}

function bytesToPoint(data) {
    if (data.length !== 65) {
        throw new Error('invalid point length');
    }
    if (data[0] !== 0x04) {
        throw new Error('only uncompressed point supported');
    }
    const xBN = new BN(data.slice(1, 33));
    const yBN = new BN(data.slice(33, 65));
    return curve.point(xBN, yBN);
}

// base64url 解码工具（JWK 使用 base64url）
function base64urlToBuffer(base64url) {
    const base64 = base64url.replace(/-/g, '+').replace(/_/g, '/');
    return Buffer.from(base64, 'base64');
}

// ---------- X.509 Base64 公钥解析 ----------
function x509Base64ToPoint(publicKeyBase64) {
    const der = Buffer.from(publicKeyBase64, 'base64');

    try {
        // 使用 Node.js 内置 crypto 解析 SPKI 公钥（DER 格式）
        const pubKey = crypto.createPublicKey({
            key: der,
            format: 'der',
            type: 'spki',
        });

        // 导出 JWK，获取 x, y 坐标 (base64url 编码)
        const jwk = pubKey.export({ format: 'jwk' });

        if (jwk.kty !== 'EC' || jwk.crv !== 'secp256k1') {
            throw new Error('not secp256k1 public key');
        }

        const xBN = new BN(base64urlToBuffer(jwk.x).toString('hex'), 16);
        const yBN = new BN(base64urlToBuffer(jwk.y).toString('hex'), 16);

        // 返回曲线点
        return curve.point(xBN, yBN);
    } catch (err) {
        // 如果 crypto 方案失败（极少数环境），可备用手工解析 DER
        // 这里提供一份简单的手动提取 BIT STRING 的方法
        // SubjectPublicKeyInfo 结构：
        // SEQUENCE { AlgorithmIdentifier, BIT STRING (04 || x || y) }
        // 简单解析：找到 BIT STRING 标记 0x03 后跳过长度字节和 unused bits 字节
        let pos = der.indexOf(0x03, 2); // 跳过开头的 SEQUENCE
        if (pos === -1) throw new Error('Invalid SPKI');

        pos += 2; // 跳过 tag 和长度
        if (der[pos] === 0x00) pos++; // unused bits 必须为 0
        const pointBytes = der.slice(pos, pos + 65); // 非压缩点 65 字节
        return bytesToPoint(pointBytes);
    }
}

// ---------- hash_to_curve (对齐 Java 版本) ----------
function hashToCurve(seed) {
    let h = sha256(seed);

    while (true) {
        const xBN = new BN(h, 'be');
        try {
            // 尝试以压缩前缀 0x02 对应的偶数 y 构造点
            const P = curve.pointFromX(xBN, false);
            return P;
        } catch (e) {
            h = sha256(h);
        }
    }
}

// ---------- 挑战码 c = H(Y || H || Gamma || U || V) mod n ----------
function hashChallenge(publicKey, h, gamma, u, v) {
    const data = Buffer.concat([
        pointToBytes(publicKey),
        pointToBytes(h),
        pointToBytes(gamma),
        pointToBytes(u),
        pointToBytes(v),
    ]);
    const hash = sha256(data);
    return new BN(hash).mod(n);
}

// ---------- 核心验证 ----------
function vrfVerify(publicKeyBase64, seed, proof, randomOutput) {
    try {
        if (!Buffer.isBuffer(proof) || proof.length !== 129) return false;
        if (!Buffer.isBuffer(randomOutput) || randomOutput.length !== 32) return false;

        // 1. 公钥
        const publicKey = x509Base64ToPoint(publicKeyBase64);

        // 2. 拆解 proof：Gamma (65) || c (32) || s (32)
        const gammaBytes = proof.slice(0, 65);
        const cBytes = proof.slice(65, 97);
        const sBytes = proof.slice(97, 129);

        const gamma = bytesToPoint(gammaBytes);
        const c = new BN(cBytes, 'be');
        const s = new BN(sBytes, 'be');

        if (c.cmp(n) >= 0 || s.cmp(n) >= 0) return false;

        // 3. H = hash_to_curve(seed)
        const h = hashToCurve(seed);

        // 4. U = s·G + (n - c)·Y
        const sG = G.mul(s);
        const negCY = publicKey.mul(n.sub(c));
        const u = sG.add(negCY);

        // 5. V = s·H + (n - c)·Gamma
        const sH = h.mul(s);
        const negCGamma = gamma.mul(n.sub(c));
        const v = sH.add(negCGamma);

        // 6. 验证挑战码
        const cPrime = hashChallenge(publicKey, h, gamma, u, v);
        if (!cPrime.eq(c)) return false;

        // 7. 验证随机输出
        const expectedOutput = sha256(pointToBytes(gamma));
        if (!expectedOutput.equals(randomOutput)) return false;

        return true;
    } catch (e) {
        return false;
    }
}

// ---------- Hex 输入包装 ----------
function vrfVerifyHex(publicKeyBase64, seedStr, proofHex, randomOutputHex) {
    const seed = Buffer.from(seedStr, 'utf8');
    const proof = Buffer.from(proofHex, 'hex');
    const randomOutput = Buffer.from(randomOutputHex, 'hex');
    return vrfVerify(publicKeyBase64, seed, proof, randomOutput);
}

// ---------- 示例 ----------
if (require.main === module) {
    const publicKeyBase64 = "publicKeyBase64";
    const seed = "seed";

    const proofHex = "proofHex";
    const randomOutputHex = "randomOutputHex";

    const ok = vrfVerifyHex(publicKeyBase64, seed, proofHex, randomOutputHex);
    console.log("verify =", ok);
}

File: VrfVerify.py
import base64
import hashlib

from ecdsa import curves, ellipticcurve
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import ec


# =========================================================
# secp256k1
# =========================================================
curve = curves.SECP256k1
G = curve.generator
n = curve.order
p = curve.curve.p()
a = curve.curve.a()
b = curve.curve.b()


# =========================================================
# 基础函数
# =========================================================
def int_to_bytes(x: int, length: int = 32) -> bytes:
    return x.to_bytes(length, "big")


def sha256(data: bytes) -> bytes:
    return hashlib.sha256(data).digest()


def point_to_bytes(P: ellipticcurve.Point) -> bytes:
    """
    非压缩格式：
    04 || x(32) || y(32)
    """
    return b"\x04" + int_to_bytes(P.x()) + int_to_bytes(P.y())


def bytes_to_point(data: bytes) -> ellipticcurve.Point:
    if len(data) != 65:
        raise ValueError("invalid point length")

    if data[0] != 0x04:
        raise ValueError("only uncompressed point supported")

    x = int.from_bytes(data[1:33], "big")
    y = int.from_bytes(data[33:65], "big")

    if not curve.curve.contains_point(x, y):
        raise ValueError("point not on curve")

    return ellipticcurve.Point(curve.curve, x, y, n)


# =========================================================
# X.509 Base64 公钥解析
# =========================================================
def x509_base64_to_point(public_key_base64: str) -> ellipticcurve.Point:
    der = base64.b64decode(public_key_base64)

    pub = serialization.load_der_public_key(der)

    if not isinstance(pub, ec.EllipticCurvePublicKey):
        raise ValueError("not EC public key")

    nums = pub.public_numbers()

    return ellipticcurve.Point(curve.curve, nums.x, nums.y, n)


# =========================================================
# hash_to_curve
# 必须与 Java 保持一致
# seed || counter(4-byte big endian)
# =========================================================
def hash_to_curve(seed: bytes) -> ellipticcurve.Point:
    """
    完全对齐 Java 版本：

    h = sha256(seed)

    while True:
        candidate = 0x02 || h

        尝试作为 secp256k1 压缩点解码
        失败则 h = sha256(h)
    """
    h = sha256(seed)

    while True:
        x = int.from_bytes(h, "big")

        # secp256k1:
        # y^2 = x^3 + 7 mod p
        rhs = (pow(x, 3, p) + 7) % p

        # 只接受偶数 y（对应压缩前缀 0x02）
        y = pow(rhs, (p + 1) // 4, p)

        if (y * y) % p == rhs:
            if y & 1:
                y = p - y

            return ellipticcurve.Point(curve.curve, x, y, n)

        h = sha256(h)


# =========================================================
# challenge
# c = H(Y || H || Gamma || U || V) mod n
# =========================================================
def hash_challenge(
    public_key: ellipticcurve.Point,
    h: ellipticcurve.Point,
    gamma: ellipticcurve.Point,
    u: ellipticcurve.Point,
    v: ellipticcurve.Point,
) -> int:
    data = (
        point_to_bytes(public_key)
        + point_to_bytes(h)
        + point_to_bytes(gamma)
        + point_to_bytes(u)
        + point_to_bytes(v)
    )

    return int.from_bytes(sha256(data), "big") % n


# =========================================================
# 核心 verify
# =========================================================
def vrf_verify(
    public_key_base64: str,
    seed: bytes,
    proof: bytes,
    random_output: bytes,
) -> bool:
    try:
        if len(proof) != 129:
            return False

        if len(random_output) != 32:
            return False

        # 公钥
        public_key = x509_base64_to_point(public_key_base64)
        #print(public_key.x(), public_key.y())

        # proof 拆解
        gamma_bytes = proof[:65]
        c_bytes = proof[65:97]
        s_bytes = proof[97:129]

        gamma = bytes_to_point(gamma_bytes)

        c = int.from_bytes(c_bytes, "big")
        s = int.from_bytes(s_bytes, "big")

        if c >= n or s >= n:
            return False

        # H = hash_to_curve(seed)
        h = hash_to_curve(seed)

        print(h.x(), h.y())

        # U' = sG - cY
        u = s * G + (n - c) * public_key

        # V' = sH - cGamma
        v = s * h + (n - c) * gamma

        # c'
        c_prime = hash_challenge(public_key, h, gamma, u, v)

        if c_prime != c:
            return False

        # random output
        expected_output = sha256(point_to_bytes(gamma))

        if expected_output != random_output:
            return False

        return True

    except Exception:
        return False


# =========================================================
# Hex 输入包装
# =========================================================
def vrf_verify_hex(
    public_key_base64: str,
    seed: str,
    proof_hex: str,
    random_output_hex: str,
) -> bool:
    proof = bytes.fromhex(proof_hex)
    random_output = bytes.fromhex(random_output_hex)

    return vrf_verify(
        public_key_base64=public_key_base64,
        seed=seed.encode("utf-8"),
        proof=proof,
        random_output=random_output,
    )


# =========================================================
# 示例
# =========================================================
if __name__ == "__main__":
    public_key_base64 = "public_key_base64"
    seed = "seed"   # detId | timestamp

    proof_hex = "proof_hex"
    random_output_hex = "random_output_hex"

    ok = vrf_verify_hex(
        public_key_base64=public_key_base64,
        seed=seed,
        proof_hex=proof_hex,
        random_output_hex=random_output_hex,
    )

    print("verify =", ok)
